synapse/storage/events.py

```diff
@@ -220,13 +221,35 @@ class EventsStore(
     EventsWorkerStore,
     BackgroundUpdateStore,
 ):
-
     def __init__(self, db_conn, hs):
         super(EventsStore, self).__init__(db_conn, hs)
 
         self._event_persist_queue = _EventPeristenceQueue()
         self._state_resolution_handler = hs.get_state_resolution_handler()
 
+        # Collect metrics on the number of forward extremities that exist.
+        self._current_forward_extremities_amount = {}
+
+        BucketCollector(
+            "synapse_forward_extremities",
+            lambda: self._current_forward_extremities_amount,
+        )
+
+        # Read the extrems every 60 sec
+        hs.get_clock().looping_call(self._read_forward_extremities, 60000)
+
+    @defer.inlineCallbacks
+    def _read_forward_extremities(self):
+        def fetch(txn):
+            txn.execute(
+                "select room_id, count(*) c from event_forward_extremities "
+                "group by room_id"
+            )
```

Comment:
(You can use mutltline string literals btw, which makes it much nicer to work with imo, as it allows easier c+p and you don't have to worry about remembering to add spaces)

---

This query takes ~25s on matrix.org from cold, so I'd vote we run it much less frequently (once an hour?).  I think if we're interested in more fine grained metrics we probably want to count the number of extremities of rooms that are active, e.g. count extremities for a room each time we do state res, count the number of extremities when we write events (we already pull out the extremities when we persist events), etc.

---

Let's also not bother pulling out the `room_id` if we're not going to use it.

Reply:
done